File: ainpc-core-plugin/src/main/java/ro/ainpc/world/patch/PatchPlannerOptions.java
package ro.ainpc.world.patch;

import java.util.List;
import java.util.Locale;
import java.util.Set;

public record PatchPlannerOptions(
    int targetPopulation,
    List<String> requiredProfessions,
    int minHouseCount,
    int maxPatchCount,
    boolean requireSocialHub,
    boolean requireQuestTriggerNode,
    Set<String> allowedCapabilities
) {
    public PatchPlannerOptions {
        targetPopulation = Math.max(0, targetPopulation);
        requiredProfessions = List.copyOf(requiredProfessions != null ? requiredProfessions : List.of());
        minHouseCount = Math.max(0, minHouseCount);
        maxPatchCount = maxPatchCount <= 0 ? 12 : maxPatchCount;
        allowedCapabilities = Set.copyOf(allowedCapabilities != null ? allowedCapabilities : defaultCapabilities());
    }

    public static PatchPlannerOptions forTargetPopulation(int targetPopulation) {
        return new PatchPlannerOptions(
            targetPopulation,
            List.of(),
            0,
            12,
            true,
            true,
            defaultCapabilities()
        );
    }

    public static PatchPlannerOptions forTargetPopulation(int targetPopulation, List<String> requiredProfessions) {
        return new PatchPlannerOptions(
            targetPopulation,
            requiredProfessions,
            0,
            12,
            true,
            true,
            defaultCapabilities()
        );
    }

    public boolean hasCapability(String capability) {
        String normalized = normalize(capability);
        return !normalized.isBlank()
            && allowedCapabilities.stream().map(PatchPlannerOptions::normalize).anyMatch(normalized::equals);
    }

    public List<String> normalizedRequiredProfessions() {
        return requiredProfessions.stream()
            .map(PatchPlannerOptions::normalize)
            .filter(value -> !value.isBlank())
            .distinct()
            .toList();
    }

    public static Set<String> defaultCapabilities() {
        return Set.of("semantic-place-mapping");
    }

    private static String normalize(String value) {
        return value == null
            ? ""
            : value.trim().toLowerCase(Locale.ROOT).replace(' ', '_').replace('-', '_');
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/world/patch/PatchPlannerOptions.kt
package ro.ainpc.world.patch

import java.util.Locale

class PatchPlannerOptions(
    targetPopulation: Int,
    requiredProfessions: List<String>?,
    minHouseCount: Int,
    maxPatchCount: Int,
    private val requireSocialHub: Boolean,
    private val requireQuestTriggerNode: Boolean,
    allowedCapabilities: Set<String>?
) {
    private val targetPopulation: Int = maxOf(0, targetPopulation)
    private val requiredProfessions: List<String> = (requiredProfessions ?: emptyList()).toList()
    private val minHouseCount: Int = maxOf(0, minHouseCount)
    private val maxPatchCount: Int = if (maxPatchCount <= 0) 12 else maxPatchCount
    private val allowedCapabilities: Set<String> = (allowedCapabilities ?: defaultCapabilities()).toSet()

    fun targetPopulation(): Int = targetPopulation
    fun requiredProfessions(): List<String> = requiredProfessions
    fun minHouseCount(): Int = minHouseCount
    fun maxPatchCount(): Int = maxPatchCount
    fun requireSocialHub(): Boolean = requireSocialHub
    fun requireQuestTriggerNode(): Boolean = requireQuestTriggerNode
    fun allowedCapabilities(): Set<String> = allowedCapabilities

    fun hasCapability(capability: String?): Boolean {
        val normalized = normalize(capability)
        return normalized.isNotBlank() &&
            allowedCapabilities.asSequence().map(::normalize).any { it == normalized }
    }

    fun normalizedRequiredProfessions(): List<String> {
        return requiredProfessions.asSequence()
            .map(::normalize)
            .filter { it.isNotBlank() }
            .distinct()
            .toList()
    }

    companion object {
        @JvmStatic
        fun forTargetPopulation(targetPopulation: Int): PatchPlannerOptions {
            return PatchPlannerOptions(
                targetPopulation,
                emptyList(),
                0,
                12,
                true,
                true,
                defaultCapabilities()
            )
        }

        @JvmStatic
        fun forTargetPopulation(targetPopulation: Int, requiredProfessions: List<String>?): PatchPlannerOptions {
            return PatchPlannerOptions(
                targetPopulation,
                requiredProfessions,
                0,
                12,
                true,
                true,
                defaultCapabilities()
            )
        }

        @JvmStatic
        fun defaultCapabilities(): Set<String> = setOf("semantic-place-mapping")

        private fun normalize(value: String?): String {
            return value?.trim()?.lowercase(Locale.ROOT)?.replace(' ', '_')?.replace('-', '_') ?: ""
        }
    }
}
